### ISSUE TITLE
Segfault on syntax checking

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11730]
Migrated from [rt.perl.org#102858](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=102858) (status was 'resolved')

Searchable as RT102858$

**Comment by p5pRT** [comment]
### From ulfer.dm@gmail.com

#### Created by ulfer.dm@gmail.com



Perl segfaults while doing syntax check with
use encoding 'utf8'\. Bug can be reproduced
with

&nbsp;   perl \-M"encoding utf8" \-e "$a ~= tr/a/b/;"

Without \-M"encoding utf8"\, it's a syntax error\,
which is ok\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.12.4:

Configured by Debian Project at Sun Oct  9 12:36:46 UTC 2011.

Summary of my perl5 (revision 5 version 12 subversion 4) configuration:

  Platform:
    osname=linux, osvers=2.6.32-5-amd64, archname=x86_64-linux-gnu-thread-multi
    uname='linux brahms 2.6.32-5-amd64 #1 smp mon oct 3 03:59:20 utc
2011 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN
-Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr
-Dprivlib=/usr/share/perl/5.12 -Darchlib=/usr/lib/perl/5.12
-Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5
-Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local
-Dsitelib=/usr/local/share/perl/5.12.4
-Dsitearch=/usr/local/lib/perl/5.12.4 -Dman1dir=/usr/share/man/man1
-Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1
-Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1
-Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh
-Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -DDEBUGGING=-g -Doptimize=-O2
-Duseshrplib -Dlibperl=libperl.so.5.12.4 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing
-pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib
/usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=, so=so, useshrplib=true, libperl=libperl.so.5.12.4
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.12.4:
    /etc/perl
    /usr/local/lib/perl/5.12.4
    /usr/local/share/perl/5.12.4
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.12
    /usr/share/perl/5.12
    /usr/local/lib/site_perl
    .


Environment for perl 5.12.4:
    HOME=/home/ulfer
    LANG=ru_RU.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From perl@profvince.com


> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> Perl segfaults while doing syntax check with
> use encoding 'utf8'\. Bug can be reproduced
> with
>
>     perl \-M"encoding utf8" \-e "$a ~= tr/a/b/;"
>
> Without \-M"encoding utf8"\, it's a syntax error\,
> which is ok\.
>

This only happens with threaded perls\. The segfault appeared somewhere 
between perl 5\.11\.0 and 5\.11\.5\.

Here's a stacktrace with perl 5\.15\.4 \(note the obviously wrong po 
argument that was passed to pad\_swipe\) :

$ gdb \-\-args perl5\.15\.4\-dbg\-thr\-64 \-M"encoding utf8" \-e "$a ~= tr/a/b/;"
GNU gdb \(Gentoo 7\.3\.1 p1\) 7\.3\.1
Copyright \(C\) 2011 Free Software Foundation\, Inc\.
License GPLv3\+&#8203;: GNU GPL version 3 or later 
\<http&#8203;://gnu\.org/licenses/gpl\.html>
This is free software&#8203;: you are free to change and redistribute it\.
There is NO WARRANTY\, to the extent permitted by law\.  Type "show copying"
and "show warranty" for details\.
This GDB was configured as "x86\_64\-pc\-linux\-gnu"\.
For bug reporting instructions\, please see&#8203;:
\<http&#8203;://bugs\.gentoo\.org/>\.\.\.
Reading symbols from 
/home/vince/perl/builds/bin/perl5\.15\.4\-dbg\-thr\-64\.\.\.done\.
\(gdb\) r
Starting program&#8203;: /home/vince/perl/builds/bin/perl5\.15\.4\-dbg\-thr\-64 
\-Mencoding\\ utf8 \-e \\ \\~=\\ tr/a/b/\\;
\[Thread debugging using libthread\_db enabled\]

Program received signal SIGSEGV\, Segmentation fault\.
0x00000000004d27b1 in Perl\_pad\_swipe \(my\_perl=0x9eb010\, po=10593312\,
&nbsp;     refadjust=true\) at pad\.c&#8203;:1524
1524        if \(PL\_curpad\[po\]\)
\(gdb\) bt
\#0  0x00000000004d27b1 in Perl\_pad\_swipe \(my\_perl=0x9eb010\, po=10593312\,
&nbsp;     refadjust=true\) at pad\.c&#8203;:1524
\#1  0x000000000041e065 in Perl\_op\_clear \(my\_perl=0x9eb010\, o=0xb29a80\)
&nbsp;     at op\.c&#8203;:649
\#2  0x000000000041dc77 in Perl\_op\_free \(my\_perl=0x9eb010\, o=0xb29a80\)
&nbsp;     at op\.c&#8203;:530
\#3  0x00000000004ca39d in Perl\_yyparse \(my\_perl=0x9eb010\, gramtype=258\)
&nbsp;     at perly\.c&#8203;:638
\#4  0x000000000045052e in S\_parse\_body \(my\_perl=0x9eb010\, env=0x0\,
&nbsp;     xsinit=0x41d192 \<xs\_init>\) at perl\.c&#8203;:2221
\#5  0x000000000044e687 in perl\_parse \(my\_perl=0x9eb010\,
&nbsp;     xsinit=0x41d192 \<xs\_init>\, argc=4\, argv=0x7fffffffde78\, env=0x0\)
&nbsp;     at perl\.c&#8203;:1628
\#6  0x000000000041d0d5 in main \(argc=4\, argv=0x7fffffffde78\,
&nbsp;     env=0x7fffffffdea0\) at perlmain\.c&#8203;:118


Vincent

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Nov 03\, 2011 at 08&#8203;:48&#8203;:16PM \+0100\, Vincent Pit wrote&#8203;:
> 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > Perl segfaults while doing syntax check with
> > use encoding 'utf8'\. Bug can be reproduced
> > with
> >
> >     perl \-M"encoding utf8" \-e "$a ~= tr/a/b/;"
> >
> > Without \-M"encoding utf8"\, it's a syntax error\,
> > which is ok\.

I think you meant

&nbsp;    perl \-M'encoding utf8' \-e '$a ~= tr/a/b/;'

in your example \(ie single quotes\) because with double quotes the shell
will interpolate $a before executing perl\. However\,

&nbsp;    perl \-M'encoding utf8' \-e '~= tr/a/b/;'

happens to be another syntax error with the same crash\.

> This only happens with threaded perls\. The segfault appeared somewhere 
> between perl 5\.11\.0 and 5\.11\.5\.

~/Perl/perl/Porting/bisect\.pl \-Dusethreads \-e 'use encoding "utf8"; eval q\{$a ~= tr/a/b/;\}'

reports\, after a lot of build output&#8203;:

0e38ac7805aa91d94a5718273a06c051670207ca is the first bad commit
commit 0e38ac7805aa91d94a5718273a06c051670207ca
Author&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
Date&#8203;:   Fri Dec 18 18&#8203;:41&#8203;:00 2009 \+0100

&nbsp;    When discarding a token free the op associated with it

&nbsp;    When discarding a token free the op associated with it\.

&nbsp;    Fixes memory leak in comp/parser\.t

&nbsp;    Gerard Goossen

&nbsp;    From 7694ec023ec8856f34964b5eeea58f1b588c89bc Mon Sep 17 00&#8203;:00&#8203;:00 2001
&nbsp;    From&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
&nbsp;    Date&#8203;: Fri\, 18 Dec 2009 18&#8203;:32&#8203;:11 \+0100
&nbsp;    Subject&#8203;: \[PATCH 4/4\] When discarding a token free the op associated with it\.
&nbsp;    Status&#8203;: O
&nbsp;    Content\-Length&#8203;: 438
&nbsp;    Lines&#8203;: 20

&nbsp;    Fixes memory leak in comp/parser\.t

&nbsp;    Signed\-off\-by&#8203;: H\.Merijn Brand \<h\.m\.brand@&#8203;xs4all\.nl>

:100644 100644 07093071e6774d0684c053ed447cc42c6f16b437 3624ca3c0424b17ddee16b037ee307f9f4aa28be M      perly\.c
bisect run success
That took 2277 seconds


I wrote the bisect tool in the hope that other people would use it too :\-/

\[Also\, that was my second attempt at a test case\. I happened to realise my
mistake stop the first before it failed\, but for illustration\, I've run it
again&#8203;:

$ ~/Perl/perl/Porting/bisect\.pl \-Dusethreads \-e 'use encoding "utf8"; $a ~= tr/a/b/;'

\.\.\.

Can't find a suitable start revision to default to\. Tried perl\-5\.002 perl\-5\.003 perl\-5\.004 perl\-5\.005 perl\-5\.6\.0 perl\-5\.8\.0 v5\.10\.0 v5\.12\.0 v5\.14\.0 at /home/nicholas/Perl/perl/Porting/bisect\.pl line 116\.
That took 571 seconds


Because\, of course\, if you don't use an eval\, the syntax error at the top
level causes a non\-zero exit\, which is seen as "bad"

This is part of why I think it's better than "just" git bisect \- it has
sanity checks to spot when your test case isn't actually testing
anything\. Just using git bisect and manually marking the start and end as
good and bad would result in git bisect \(correctly\) telling you that the
child of the start commit was the first bad commit\. After wasting a lot of
time\.

Also helping laziness is that it defaults to trying to find a start revision
for you\, which saves you having to work out when the thing your want to test
became valid syntax\.\]

Nicholas Clark

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Nov 04 06&#8203;:13&#8203;:26 2011\, nicholas wrote&#8203;:
> I wrote the bisect tool in the hope that other people would use it too
> :\-/

I used it\! Once\, that is\, resulting in
\<https://rt.cpan.org/Ticket/Display.html?id=72158>.

But I find it awfully hard to figure out how to install modules while
running it and how to find the commit that caused a warning to start
appearing\.  If you could make it more like Andreas König’s
binsearchaperl\, it would be even easier to use\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Fri Nov 04 06&#8203;:13&#8203;:26 2011\, nicholas wrote&#8203;:
> I wrote the bisect tool in the hope that other people would use it too
> :\-/

I used it\! Once\, that is\, resulting in
\<https://rt.cpan.org/Ticket/Display.html?id=72158>.

But I find it awfully hard to figure out how to install modules while
running it and how to find the commit that caused a warning to start
appearing\.  If you could make it more like Andreas König’s
binsearchaperl\, it would be even easier to use\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Nov 03 12&#8203;:48&#8203;:53 2011\, perl@&#8203;profvince\.com wrote&#8203;:
> 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> >
> > Perl segfaults while doing syntax check with
> > use encoding 'utf8'\. Bug can be reproduced
> > with
> >
> >     perl \-M"encoding utf8" \-e "$a ~= tr/a/b/;"
> >
> > Without \-M"encoding utf8"\, it's a syntax error\,
> > which is ok\.
> >
> 
> This only happens with threaded perls\. The segfault appeared somewhere 
> between perl 5\.11\.0 and 5\.11\.5\.

Since it’s a regression\, I’ve made it a blocker for 5\.16\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Fixed with commit 9100eeb18\.

\-zefram

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
